Misc. object_wrapper issues: testing, refactoring

jdavcs opened the pull request:
This is a draft.
Ref #12358

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jdavcs committed:
Add first unit tests for object_wrapper, refactor

jdavcs committed:
Refactor: remove redundant check

SafeStringWrapper is always in no_wrap_classes

jdavcs committed:
Refactor tests (use fixture, make more readable)

jdavcs committed:
Fix bug: incorrect import; add test (see note)

sqlalchemy.orm.state.InstanceState is a "semi-public" class (as per SA
docs). Accessing it using its fully qualified name may raise an error
because it may have not been preloaded (see https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/__init__.py)
(Exposed when running the test as a standalone).

The added test tests the usecase addressed by InstanceState.

jdavcs committed:
Unit tests for object_wrapper should be in security

jdavcs committed:
Add characterization test to verify const value

Trivial, but given the critical function of this module, this is a
reasonable precaution, I think. Better safe than sorry.

jmchilton commented on test/unit/security/test_object_wrapper.py:
```diff
@@ -0,0 +1,88 @@
+import pytest
```
Can you place this unit test in test/unit/data/security/test_object_wrapper.py instead.

The top-level of test/unit should be directories corresponding to the packages the code belongs in (this is aspirational I know 😅 but it is what I've been working toward). If you place this file in there, it will automatically included in the CI tests for galaxy-data.

jdavcs replied:
Of course! Will do.
> The top-level of test/unit should be directories corresponding to the packages the code belongs in 

This makes perfect sense.

jdavcs committed:
Move unit test into unit/data/securiry

jdavcs committed:
Rename util.valid_chars, add test

This will be imported by other modules, so make it clear it's a
constant.

jdavcs committed:
Rename util.mapped_chars, add test

Will be imported by other modules, needs to appear a constant

jdavcs committed:
Remove refs to char_map, invalid_chars from object_wrapper

This is all handled in util/__init__.py

jdavcs committed:
Refactor usage of INVALID_CHARACTER, add test

jdavcs committed:
Remove more duplicate usage of INVALID_CHARACTER

jdavcs committed:
Remove duplicate use of VALID_CHARACTERS

jdavcs committed:
Add import (used as str in eval()), ignore flake8 error

jdavcs committed:
Use modifed versions of valid/mapped chars in object_wrapper

jdavcs committed:
Remove BufferType, DictProxyType; rename XRangeType

(Minor Python 3 cleanup)

jdavcs committed:
Test wrap-no-subclass + verify const value

jdavcs committed:
Test assigniment of Callable- vs SafeStringWrapper

jdavcs committed:
Test wrap sequences, mapping, simplify tests, code

jdavcs committed:
Rearrange tests

jdavcs committed:
Harden the test for safe_name: use a more general case

jdavcs committed:
Fix wrapped_class_name bug (see note)

At this point in the code, we may have a CallableSafeStringWrapper. This
is unlikely, but we provide for this possiblity by referring to the
class using the safe_class variable when creating the new type a few
lines below.

An unlikely use case of a CallableSafeStringWrapper is covered by a unit
test (instance of a class that has the __call__() method defined)

jdavcs committed:
Factor out determining wrapped class; add test

+ Rename local variable: class_to_wrap != wrapped_class (simplifies
  reasoning about the code, I think)

+ Fix error in test_safe_class: we are passing instances, not their
  classes, to the wrapper

jdavcs committed:
Test for wrapper class reuse

jdavcs committed:
Add unwrap function + test

jdavcs committed:
Refactor comparison methods, add tests

jdavcs committed:
Drop __cmp__ method: dropped in Python 3.0

jdavcs committed:
Add tests for trivial methods

jdavcs committed:
Fix infinite loop bug (see note)

The while loop does nothing for the return value. However, if
self.unsanitized happens to be an instance of SafeStringWrapper, the
loop will not terminate.

jdavcs committed:
Add tests for numerical methods; add IntWrapper

jdavcs committed:
Drop __div__ method: dropped in Python 3

jdavcs committed:
Refactor numerical methods

jdavcs committed:
Test SafeStringWrapper constants

jdavcs committed:
Remove Python2 stuff: cmp, coerse, nonzero

jdavcs committed:
Remove slice methods removed in Python 3

Ref: https://docs.python.org/3/whatsnew/3.0.html?highlight=__getslice__#operators-and-special-methods

jdavcs committed:
Add test for str and repr methods

jdavcs committed:
Improve IntWrapper for tests; include most methods from data model

jdavcs committed:
Remove __oct__ and __hex__ removed in Python 3

jdavcs committed:
Test return values of overridden numeric methods;

Include tests for r* methods (swapped operands): to be included in
SafeStringWrapper in subsequent commits

jdavcs committed:
Add IllegalReturnValueSimulator for testing

jdavcs committed:
Add tests for sanitized return value of bin arithmetic

jdavcs committed:
Add tests for bin arithmetic w/swapped operands

Expect to fail: methods not added yet (split into 2 commits for clarity)

jdavcs committed:
Add r* numeric method overrides; refactor and/or/xor/rpow

jdavcs committed:
Add tests for unary arithmetic methods

When overriding, don't use operators: call methods directly for
clarity (we are delegating to the same method on self.unsanitized)

jdavcs committed:
Add misc numeric methods, tests (see note)

See inline comment in test module on testing these methods.

jdavcs committed:
Fix bug: call __index__(), not index()

violethaze74 reviewed (commented):
reviewed
